### ISSUE TITLE
amp-story: Pages visually break when navigating with keyboard

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#32506]
# Pages visually break when navigating with keyboard

## WCAG Level
Level A

## Priority
Critical

## Pages/screens/components affected
* [Navigation](https://www.gstatic.com/amphtml/stamp/qa/navigation.html)
* [Sharing](https://www.gstatic.com/amphtml/stamp/qa/sharing.html)
* [Tooltip](https://www.gstatic.com/amphtml/stamp/qa/tooltip.html)
* [AMP Story Player - Player](https://www.gstatic.com/amphtml/stamp/qa/player.html)

## Description
Good design means supporting different modes of interaction, because there are many reasons why someone might use a keyboard instead of a mouse, and if keyboard interaction is not fully supported the website is effectively unusable.

Some of the tested stories visually "break" (with misaligned pages) when navigating using the keyboard and the `Tab` key, particularly in the small-screen/smartphone view, making them unreadable/unusable.

In some cases, the visual break is only minor - for instance, in [Navigation](https://www.gstatic.com/amphtml/stamp/qa/navigation.html) and [Tooltip](https://www.gstatic.com/amphtml/stamp/qa/tooltip.html), in large-screen/desktop view, the first `Tab` just results in the layout (including progress indicators) shifting to the left of the screen.

![Illustration of the layout moving to the left on first Tab in the navigation sample](https://user-images.githubusercontent.com/54933393/107247896-adfa5680-6a29-11eb-9c89-1a560a44a84f.png)

![The layout shifted to the left on first Tab in the tooltip sample](https://user-images.githubusercontent.com/54933393/107247997-c8343480-6a29-11eb-9b5c-2ca03d8f1e1b.png)

However, in small-screen/smartphone view, `Tab` navigation can result in very broken layout, with slides partially moved out of view, and multiple slides showing partially at the same time.

![Example of broken layout in small-screen/smartphone view](https://user-images.githubusercontent.com/54933393/107248052-d71ae700-6a29-11eb-8c70-24de147d7dad.png)

![Example of broken layout in small-screen/smartphone view](https://user-images.githubusercontent.com/54933393/107248122-e6019980-6a29-11eb-861b-019710151705.png)

![Example of broken layout in small-screen/smartphone view](https://user-images.githubusercontent.com/54933393/107248176-f0bc2e80-6a29-11eb-9281-1fe583cda63a.png)

![Example of broken layout in small-screen/smartphone view](https://user-images.githubusercontent.com/54933393/107248215-fc0f5a00-6a29-11eb-8873-324ba93a41e4.png)

In [Sharing](https://www.gstatic.com/amphtml/stamp/qa/sharing.html), in large-screen/desktop view, navigating past the "next page" control with `Tab` results in a completely blank screen.

In small-screen/desktop view, after opening the "Sharing" dialog, using `Tab` can result in both the sharing dialog and the "bookend" dialog showing simultaneously.

![small-screen/smartphone view of the sharing story, with both the sharing dialog and bookend visible at the bottom of the screen](https://user-images.githubusercontent.com/54933393/107248269-0893b280-6a2a-11eb-9272-a23123394bcb.png)

## User impact
Someone with mobility problems will often use a keyboard instead of a mouse to navigate a web page. A screen reader user will also tend to use a keyboard if they cannot see the screen to use the mouse. Therefore if the website does not properly support keyboard interaction a large number of people are effectively prevented from using it.

In this case, keyboard users will find it difficult or impossible to navigate through certain AMP stories, as the content of pages will simply disappear/become badly misaligned.

It's worth emphasising that the small-screen/smartphone view of AMP Stories won't only be used on an actual touch screen operated smartphone. This view will also be used by desktop users that have sized/zoomed their browser window to a small-enough size to trigger the "smartphone" view, but are still navigating using keyboard or mouse. And even on actual smartphones, users may be navigating with an external mouse or keyboard. For this reason it's crucial that this view also work well for keyboard users.

## Required solution
Review how the layout for small-screen/smartphone view is currently implemented, and make sure that pages don't visually break when users navigate using the keyboard.

Part of the solution is likely tied to how focus order / focusable elements are handled when they are not part of the currently active/visible page (see separate issue on focus order), combined with the specific styles used to visually style and center each page.

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Check that all interactive elements can be navigated to using the tab key.
2. Check that all interactive elements can be activated using the enter or space keys.
3. Check that interactive elements can be interacted with using the arrow keys.

### Definition of done
Complete all of these tasks before closing this issue or indicating it is ready for retest:
* All issues identified within the test sample have been resolved.
* The rest of the website has been tested for the same issue.
* All issues identified throughout the rest of the website have been resolved or filed as new issues.

## Related standards
* [WCAG 2.1 Success Criterion 2.1.1 Keyboard (Level A)](https://www.w3.org/TR/WCAG21/#keyboard)

## More information
* [WebAIM: Keyboard](https://webaim.org/techniques/keyboard/)

## Test data
Test date: January 2021

**Comment by processprocess** [comment]
Duplicate of #32590 which provides some context and proposals.

Context/nuances on tabbing:
https://github.com/ampproject/amphtml/issues/31747#issuecomment-754734271

**Comment by processprocess** [comment]
List of other [problematic tubbable components](https://github.com/ampproject/amphtml/issues/31747#issuecomment-784496735) in stories.

**Comment by mszylkowski** [comment]
PRs like #32859 and #32840 should fix most of these issues.

**Comment by gmajoulet** [comment]
@processprocess @mszylkowski I believe this can be closed?

**Comment by mszylkowski** [comment]
Closing this since amp-twitter was handled on #33919, and the original issue was closed
